common/archiver/gcloud/queryParser.go

```diff
@@ -0,0 +1,306 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source queryParser.go -destination queryParser_mock.go -mock_names Interface=MockQueryParser
+
+package gcloud
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/xwb1989/sqlparser"
+
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+)
+
+type (
+	// QueryParser parses a limited SQL where clause into a struct
+	QueryParser interface {
+		Parse(query string) (*parsedQuery, error)
+	}
+
+	queryParser struct{}
+
+	parsedQuery struct {
+		workflowID      *string
+		workflowType    *string
+		startTime       int64
+		closeTime       int64
+		searchPrecision *string
+		runID           *string
+		emptyResult     bool
+	}
+)
+
+// All allowed fields for filtering
+const (
+	WorkflowID      = "WorkflowID"
+	RunID           = "RunID"
+	WorkflowType    = "WorkflowType"
+	CloseTime       = "CloseTime"
+	StartTime       = "StartTime"
+	CloseStatus     = "CloseStatus"
+	SearchPrecision = "SearchPrecision"
+)
+
+// Precision specific values
+const (
+	PrecisionDay    = "Day"
+	PrecisionHour   = "Hour"
+	PrecisionMinute = "Minute"
+	PrecisionSecond = "Second"
+)
+
+const (
+	queryTemplate = "select * from dummy where %s"
+
+	defaultDateTimeFormat = time.RFC3339
+)
+
+// NewQueryParser creates a new query parser for filestore
+func NewQueryParser() QueryParser {
+	return &queryParser{}
+}
+
+func (p *queryParser) Parse(query string) (*parsedQuery, error) {
+	stmt, err := sqlparser.Parse(fmt.Sprintf(queryTemplate, query))
+	if err != nil {
+		return nil, err
+	}
+	whereExpr := stmt.(*sqlparser.Select).Where.Expr
+	parsedQuery := &parsedQuery{}
+	if err := p.convertWhereExpr(whereExpr, parsedQuery); err != nil {
+		return nil, err
+	}
+
+	if parsedQuery.workflowType == nil {
+		parsedQuery.workflowType = new(string)
+	}
+
+	if parsedQuery.workflowID == nil {
+		parsedQuery.workflowID = new(string)
+	}
```

Comment:
This looks a little bit weird to me, how do you tell the difference between `workflowType is not specified` and `workflowType should be empty string` (if user has a typo)? I think a better way is move those checks to the `query()` function. Check if the fields are nil or not. If not, create a preCondition based on the value and pass the condition to query client.

Same for the `runID` field below.